### PR TITLE
Unify subsampling with stratify

### PR DESCRIPTION
**Pull request by isaac-chung**
Addresses https://github.com/embeddings-benchmark/mteb/issues/508

Also added a rerun as a comparison.

### CODE REVIEW

**Review by imenelydiaker — APPROVED** [review]
This looks great, thank you !

**Review by imenelydiaker — COMMENTED** [review]
Just one thing, should we allow subsampling without stratification? For datasets that don't have a label column for example?

**Comment by isaac-chung** [comment]
> Just one thing, should we allow subsampling without stratification? For datasets that don't have a label column for example?

I'd say this base case is covered by
```
self.dataset["test"] = (
    self.dataset["test"].shuffle(seed=self.seed).select(range(TEST_SAMPLES))
)
```

wdyt?

**Comment by imenelydiaker** [comment]
> > Just one thing, should we allow subsampling without stratification? For datasets that don't have a label column for example?
> 
> I'd say this base case is covered by
> 
> ```
> self.dataset["test"] = (
>     self.dataset["test"].shuffle(seed=self.seed).select(range(TEST_SAMPLES))
> )
> ```
> 
> wdyt?

Yep true!

**Comment by KennethEnevoldsen** [comment]
@isaac-chung when I doubt just use fix it bumps the version (which we basically want to do every time we change the codebase)

**Comment by isaac-chung** [comment]
Will do next time. Thanks!

File: mteb/abstasks/AbsTask.py [hunk 1]
```diff
@@ -12,6 +13,7 @@
 
 class AbsTask(ABC):
     metadata: TaskMetadata
+    max_n_samples = 2048
```
Comment (KennethEnevoldsen) [inline mteb/abstasks/AbsTask.py:16]: @isaac-chung I believe this line is problematic as it can be understood as the max number of samples for the dataset (which is not actually enforced). I would instead move it to `stratified_subsampling` as an argument with a default.

File: mteb/abstasks/AbsTask.py [hunk 2]
```diff
@@ -32,6 +34,31 @@ def dataset_transform(self):
         """
         pass
 
+    def stratified_subsampling(
+        self,
+        splits: Union[str, list[str]] = ["test"],
```
Comment (KennethEnevoldsen) [inline mteb/abstasks/AbsTask.py:39]: I would just use list[str] (following the mantra: `There should be one-- and preferably only one --obvious way to do it.`)

File: mteb/abstasks/AbsTask.py [hunk 3]
```diff
@@ -32,6 +34,31 @@ def dataset_transform(self):
...
+        label: str = "label",
+        n_samples: int = max_n_samples,
+    ):
+        """Subsamples the dataset with stratification by the supplied label.
+        The following kwargs must be provided
+        for stratified_subsampling to run:
+        - splits: Union[str, list[str]], the splits of the dataset.
+        - label: str, the label with which the stratified sampling is based on.
+        - n_samples: Optional[int], number of samples to subsample. Default is max_n_samples.
+        """
```
Comment (KennethEnevoldsen) [inline mteb/abstasks/AbsTask.py:49]: reformat to google style docstrings

File: mteb/abstasks/AbsTask.py [hunk 4]
```diff
@@ -32,6 +34,31 @@ def dataset_transform(self):
...
+            splits = [splits]
+
+        ## Can only do this if the label column is of ClassLabel.
+        if not isinstance(self.dataset[splits[0]].features[label], datasets.ClassLabel):
+            self.dataset = self.dataset.class_encode_column(label)
+
+        for split in splits:
+            self.dataset[split] = self.dataset[split].train_test_split(
+                test_size=n_samples, seed=self.seed, stratify_by_column=label
+            )["test"]  ## only take the specified test split.
```
Comment (KennethEnevoldsen) [inline mteb/abstasks/AbsTask.py:60]: I would instead create the dataset and return it and convert to method to a @staticmethod. This avoids state which can often lead to bugs.

**Review by KennethEnevoldsen — COMMENTED** [review]
Good thing that you guys merged this in (feel free to do that in the future as well), I have adressed a few issues -  @isaac-chung will you do these in a new PR?

**Comment by isaac-chung** [comment]
PR coming right up.
